File: TEMA2/Actividad_2/ZIG/collatz.zig
const std = @import("std");

// Función pura: matemática estricta para la conjetura de Collatz
fn collatz(n: u64) u64 {
    var count: u64 = 0;
    var current = n;

    while (current != 1) {
        if (current % 2 == 0) {
            current /= 2;
        } else {
            current = current * 3 + 1;
        }
        count += 1;
    }

    return count;
}

pub fn main() void {
    std.debug.print("[Zig] Iniciando procesamiento intensivo (Collatz)...\n", .{});

    // Límite establecido para la prueba de carga
    const limite_n: u64 = 10_000_000;
    var max_secuencia: u64 = 0;

    // Bucle principal de procesamiento
    var i: u64 = 1;
    while (i < limite_n) : (i += 1) {
        const longitud = collatz(i);
        if (longitud > max_secuencia) {
            max_secuencia = longitud;
        }
    }

    std.debug.print("✅ Procesamiento completado.\n", .{});
    std.debug.print("🔢 Maxima longitud de secuencia encontrada: {}\n", .{max_secuencia});
}

File: TEMA2/Actividad_2/RUST/collatz.rs
use std::time::Instant;

// Función pura: usamos mutabilidad explícita (mut) solo donde se requiere
fn collatz(mut n: u64) -> u64 {
    let mut count: u64 = 0;
    
    while n > 1 {
        if n % 2 == 0 {
            n /= 2;
        } else {
            n = n * 3 + 1;
        }
        count += 1;
    }
    
    // Retorno idiomático en Rust (sin 'return' ni ';')
    count
}

fn main() {
    println!("[Rust] Iniciando procesamiento intensivo (Collatz)...");

    const LIMITE_N: u64 = 10_000_000;
    let mut max_secuencia: u64 = 0;

    // Cronómetro nativo de Rust
    let start = Instant::now();

    // Bucle iterativo utilizando rangos inclusivos (1..=LIMITE_N)
    for i in 1..=LIMITE_N {
        let longitud = collatz(i);
        if longitud > max_secuencia {
            max_secuencia = longitud;
        }
    }

    let duration = start.elapsed();
    // Convertimos la duración a milisegundos flotantes para mayor precisión
    let elapsed_ms = duration.as_secs_f64() * 1000.0;

    println!("✅ Procesamiento completado.");
    println!("⏱️  Tiempo de ejecución interno: {:.2} ms", elapsed_ms);
    println!("🔢 Máxima longitud de secuencia encontrada: {}", max_secuencia);
}